### PR TITLE
Workout

### PR DESCRIPTION
**Pull request by craig-a-lee**
- User now able to logout via settings page. 
- Refined user profile/library UI design. 
- "Liked Exercises" workout is created each time a user signs up for the app. This workout can be viewed on profile tab. 
- Added Workout, WorkoutCell, SettingsViewController classes

Some comments may be seen but code will be refined tomorrow.

### CODE REVIEW

File: Wellthier/API/ExerciseAPIManager.m [hunk 1]
```diff
@@ -51,4 +47,33 @@ - (void) fetchAllExercises: (void(^)(NSArray *exercises, NSError *error))complet
     [dataTask resume];
 }
 
+- (NSMutableURLRequest *) getHelper: (NSString *) itemsName {
```
Comment (vijaykanse) [inline Wellthier/API/ExerciseAPIManager.m]: May be rename it to urlRequestForItems:(NSString *)itemsName ?

File: Wellthier/Models/SearchCell.m [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+//  SearchCell.m
+//  Wellthier
+//
+//  Created by Craig Lee on 7/11/22.
+//
+
+#import "SearchCell.h"
+
+@implementation SearchCell
+- (IBAction)buttonPressed:(id)sender {
```
Comment (vijaykanse) [inline Wellthier/Models/SearchCell.m]: remove action if not needed.

File: Wellthier/View Controllers/GifViewController.m [hunk 3]
```diff
@@ -24,23 +25,17 @@ - (void)viewDidLoad {
 
 - (void) setParams:(Exercise *)exercise {
     AnimatedGif *newGif = [AnimatedGif getAnimationForGifAtUrl:exercise.gifUrl];
-    [self.gif setAnimatedGif:newGif startImmediately:YES];
+    [self.gifImageView setAnimatedGif:newGif startImmediately:YES];
     NSLog(@"here: %@", exercise.gifUrl);
```
Comment (vijaykanse) [inline Wellthier/View Controllers/GifViewController.m]: remove NSLog

File: Wellthier/View Controllers/LibraryViewController.h [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+//  Created by Craig Lee on 7/13/22.
+//
+
+#import "ViewController.h"
+@import Parse;
+NS_ASSUME_NONNULL_BEGIN
+
+@interface LibraryViewController : ViewController
+@property (weak, nonatomic) IBOutlet PFImageView *profilePic;
+@property (strong, nonatomic) PFUser *currUser;
```
Comment (vijaykanse) [inline Wellthier/View Controllers/LibraryViewController.h]: rename it to currentUser ?

File: Wellthier/View Controllers/LibraryViewController.m [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+/*
+#pragma mark - Navigation
+
+// In a storyboard-based application, you will often want to do a little preparation before navigation
+- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
+    // Get the new view controller using [segue destinationViewController].
+    // Pass the selected object to the new view controller.
+}
+*/
```
Comment (vijaykanse) [inline Wellthier/View Controllers/LibraryViewController.m]: remove

File: Wellthier/View Controllers/LoginViewController.h [hunk 6]
```diff
@@ -13,6 +13,8 @@ NS_ASSUME_NONNULL_BEGIN
 @property (weak, nonatomic) IBOutlet UITextField *usernameField;
 @property (weak, nonatomic) IBOutlet UITextField *passwordField;
 
+- (IBAction)gesture:(id)sender;
```
Comment (vijaykanse) [inline Wellthier/View Controllers/LoginViewController.h]: what is this gesture about ? Can we rename  and give contextual name ?

File: Wellthier/View Controllers/NewWorkoutViewController.m [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+/*
+#pragma mark - Navigation
+
+// In a storyboard-based application, you will often want to do a little preparation before navigation
+- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
+    // Get the new view controller using [segue destinationViewController].
+    // Pass the selected object to the new view controller.
+}
+*/
```
Comment (vijaykanse) [inline Wellthier/View Controllers/NewWorkoutViewController.m]: remove

File: Wellthier/View Controllers/SearchViewController.h [hunk 8]
```diff
@@ -6,17 +6,23 @@
...
 @property (weak, nonatomic) IBOutlet UIButton *nameButton;
 @property (weak, nonatomic) IBOutlet UIButton *targetMuscleButton;
 @property (weak, nonatomic) IBOutlet UIButton *bodyPartButton;
 @property (weak, nonatomic) IBOutlet UIButton *equipmentButton;
 @property (weak, nonatomic) IBOutlet UISearchBar *searchBar;
-@property (nonatomic, strong) NSMutableArray *arrayOfExercises;
-@property (nonatomic, strong) NSMutableArray *filteredExercises;
+@property (nonatomic, strong) NSMutableArray <Exercise *> *arrayOfExercises;
+@property (nonatomic, strong) NSMutableArray <Exercise *> *filteredExercises;
+@property (nonatomic, strong) NSArray <NSString *> *bodyParts;
```
Comment (vijaykanse) [inline Wellthier/View Controllers/SearchViewController.h:22]: Awesome, would recommend to follow this pattern in other parts of app

File: Wellthier/View Controllers/SettingsViewController.m [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+/*
+#pragma mark - Navigation
+
+// In a storyboard-based application, you will often want to do a little preparation before navigation
+- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
+    // Get the new view controller using [segue destinationViewController].
+    // Pass the selected object to the new view controller.
+}
+*/
```
Comment (vijaykanse) [inline Wellthier/View Controllers/SettingsViewController.m]: remove

**Review by vijaykanse — APPROVED** [review]

File: Wellthier/Models/Workout.h [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+@interface Workout : PFObject <PFSubclassing>
+
+@property (nonatomic, strong) NSString *postID;
+@property (nonatomic, strong) NSString *userID;
+@property (nonatomic, strong) PFUser *author;
+
+@property (nonatomic, strong) NSString *title;
+@property (nonatomic, strong) PFFileObject *image;
+@property (nonatomic, strong) NSNumber *likeCount;
+@property (nonatomic, strong) NSMutableArray *exercises;
```
Comment (xueyike) [inline Wellthier/Models/Workout.h]: Let's reorder the params and list the same type together.

File: Wellthier/Models/Workout.m [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+}
+
++ (void) postUserWorkout: ( UIImage * _Nullable )image withTitle: ( NSString * _Nullable )title withCompletion: (PFBooleanResultBlock  _Nullable)completion {
+        Workout *newWorkout = [Workout new];
+        newWorkout.image = [self getPFFileFromImage:image];
+        newWorkout.author = [PFUser currentUser];
+        newWorkout.title = title;
+        newWorkout.exercises = [NSMutableArray new];
+        newWorkout.likeCount = @(0);
+
```
Comment (xueyike) [inline Wellthier/Models/Workout.m]: remove extra empty line.

File: Wellthier/Models/Workout.m [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+        newWorkout.title = title;
+        newWorkout.exercises = [NSMutableArray new];
+        newWorkout.likeCount = @(0);
+    
+        
+        [newWorkout saveInBackgroundWithBlock: completion];
+}
+
++ (PFFileObject *)getPFFileFromImage: (UIImage * _Nullable)image {
+
```
Comment (xueyike) [inline Wellthier/Models/Workout.m]: remove extra empty line.

File: Wellthier/Models/Workout.m [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+    NSData *imageData = UIImagePNGRepresentation(image);
+    // get image data and check if that is not nil
+    if (!imageData) {
+        return nil;
+    }
+    
+    return [PFFileObject fileObjectWithName:@"image.png" data:imageData];
+}
+    
+
```
Comment (xueyike) [inline Wellthier/Models/Workout.m:50]: remove extra empty line.

File: Wellthier/Models/WorkoutCell.m [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+//
+//  Created by Craig Lee on 7/13/22.
+//
+
+#import "WorkoutCell.h"
+#import "Workout.h"
+
+@implementation WorkoutCell
+
+- (void)setParams:(Workout *)detailWorkout {
```
Comment (xueyike) [inline Wellthier/Models/WorkoutCell.m:13]: You are doing setParam, reload background and update content in the same method. the "setParams:" name may not be appropriate. Maybe name it "prepareCell:" instead?

File: Wellthier/View Controllers/NewWorkoutViewController.m [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+    
+    UIAlertAction *chooseAction = [UIAlertAction actionWithTitle:@"Choose Picture"
+                                                       style:UIAlertActionStyleDefault
+                                                     handler:^(UIAlertAction * _Nonnull action) {
+                                                             // handle response here.
+        if ([UIImagePickerController isSourceTypeAvailable:UIImagePickerControllerSourceTypeCamera]) {
+            imagePickerVC.sourceType = UIImagePickerControllerSourceTypeCamera;
+        }
+        else {
+            NSLog(@"Camera 🚫 available so we will use photo library instead");
```
Comment (xueyike) [inline Wellthier/View Controllers/NewWorkoutViewController.m:37]: W should provide choose from album option regardless of the availability of camera. The "Camera 🚫 available" text is inappropriate.

**Pull request merged**